src/storage/storage.c

```diff
@@ -343,7 +351,7 @@ storageInfoListSort(
         // If no sorting then use the callback directly
         if (sortOrder == sortOrderNone)
         {
-            result = storageInterfaceInfoListP(this->driver, path, callback, callbackData);
+            result = storageInterfaceInfoListP(this->driver, path, level, callback, callbackData, .expression = expression);
```

Comment:
Use of the expression by the driver is meant to be optional and used only if the driver can avoid some expensive disk/network IO. This has now been documented in `storage.intern.h`.

That said, the S3 driver should not have been doing the full regexp check. It should only be using the prefix to limit the results that come back from S3.

The overall idea is to move this check up to the storage interface so the drivers don't need to worry about it unless they have a good reason to. The potential extra callbacks are not a big worry since the IO needed to get the list is many orders of magnitude more expensive.

Reply:
Also added this to the PR description which will become the commit comment.